### PR TITLE
Replace `room_invite_state_types` with `room_prejoin_state`

### PR DESCRIPTION
**Pull request by richvdh**
`room_invite_state_types` was inconvenient as a configuration setting, because
anyone that ever set it would not receive any new types that were added to the
defaults. Here, we deprecate the old setting, and replace it with a couple of
new settings under `room_prejoin_state`.

This is part of fixing #9448.

### CODE REVIEW

**Review by clokep — APPROVED** [review]
Seems fine if CI passes!

This will conflict / have some ramifications for the knocking implementation in #6739.

File: synapse/config/api.py [hunk 1]
```diff
@@ -12,38 +12,127 @@
...
+           #
+           # Uncomment the following to disable these defaults (so that only the event
+           # types listed in 'additional_event_types' are shared). Defaults to 'false'.
+           #
+           #disable_default_event_types: true
+
+           # Additional state event types to share with users when they are invited
+           # to a room.
+           #
+           # By default, this list is empty (so only the default event types are shared)
```
Comment (anoadragon453) [inline synapse/config/api.py:58]:
```suggestion
           # By default, this list is empty (so only the default event types are shared).
```

File: synapse/config/api.py [hunk 2]
```diff
@@ -12,38 +12,127 @@
...
+        if not room_prejoin_state_config.get("disable_default_event_types"):
+            yield from _DEFAULT_PREJOIN_STATE_TYPES
+
+        yield from room_prejoin_state_config.get("additional_event_types", [])
+
+
+_ROOM_INVITE_STATE_TYPES_WARNING = """\
+WARNING: The 'room_invite_state_types' configuration setting is now deprecated,
+and replaced with 'room_prejoin_state'. New features may not work correctly
+unless room_invite_state_types is removed. See the sample configuration file for
```
Comment (anoadragon453) [inline synapse/config/api.py:97]:
```suggestion
unless 'room_invite_state_types' is removed. See the sample configuration file for
```

File: synapse/handlers/message.py [hunk 3]
```diff
@@ -385,7 +385,7 @@ def __init__(self, hs: "HomeServer"):
         self._events_shard_config = self.config.worker.events_shard_config
         self._instance_name = hs.get_instance_name()
 
-        self.room_invite_state_types = self.hs.config.room_invite_state_types
+        self.room_invite_state_types = self.hs.config.api.room_prejoin_state
```
Comment (anoadragon453) [inline synapse/handlers/message.py:388]: Is this preferred over `self.hs.config.room_prejoin_state` for performance reasons?

File: synapse/config/api.py [hunk 4]
```diff
@@ -12,38 +12,127 @@
...
+           #additional_event_types:
+           #  - org.example.custom.event.type
+        """ % {
+            "formatted_default_state_types": formatted_default_state_types
+        }
+
+    def _get_prejoin_state_types(self, config: JsonDict) -> Iterable[str]:
+        """Get the event types to include in the prejoin state
+
+        Parsees the config and returns an iterable of the event types to be included.
```
Comment (anoadragon453) [inline synapse/config/api.py:69]:
```suggestion
        Parses the config and returns an iterable of the event types to be included.
```

File: synapse/handlers/message.py (hunk 3 above)
Comment (richvdh) [reply synapse/handlers/message.py:388]: in my ideal world, we'd get rid of all the direct accesses to `config.foo` fields, and remove the magic in `RootConfig` that makes it work.

In the meantime, I'm working to reduce instances of it, one by one.

**Comment by richvdh** [comment]
thanks @anoadragon453 for your attention to detail!

**Review by anoadragon453 — APPROVED** [review]

**Comment by rubo77** [comment]
which is the first matrix version, where you have to change this to room_prejoin_state?

**Comment by clokep** [comment]
> which is the first matrix version, where you have to change this to room_prejoin_state?

This option is first available in Synapse v1.32.0, but the code should be backwards compatible. If you're seeing a bug please file a new issue!

**Pull request merged**